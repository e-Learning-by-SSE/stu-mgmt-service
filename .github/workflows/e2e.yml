name: E2E Tests
on: push

jobs:
  # Label of the runner job
  runner-job:
    # You must use a Linux environment when using service containers or container jobs
    runs-on: ubuntu-latest

    # Service containers to run with `runner-job`
    services:
      # Label used to access the service container
      postgres:
        # Docker Hub image
        image: postgres:latest
        env:
          POSTGRES_DB: StudentMgmtDb
          POSTGRES_USER: postgres
          POSTGRES_PASSWORD: admin
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 5432:5432

    steps:
      # Downloads a copy of the code in your repository before running CI tests
      - name: Check out repository code
        uses: actions/checkout@v2

      # Performs a clean installation of all dependencies in the `package.json` file
      # For more information, see https://docs.npmjs.com/cli/ci.html
      - name: Install dependencies
        run: npm ci

      - name: Run e2e tests
        run: npm run test:e2e
