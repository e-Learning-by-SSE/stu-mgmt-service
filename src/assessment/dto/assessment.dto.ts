import { ApiProperty, ApiPropertyOptional, OmitType } from "@nestjs/swagger";
import { AssignmentDto } from "../../course/dto/assignment/assignment.dto";
import { ParticipantDto } from "../../course/dto/course-participant/participant.dto";
import { GroupDto } from "../../course/dto/group/group.dto";
import { UserDto } from "../../shared/dto/user.dto";
import { UserId } from "../../shared/entities/user.entity";
import { PartialAssessmentDto } from "./partial-assessment.dto";

export class AssessmentDto {
	/** Unique identifier of this assessment. */
	@ApiProperty({ description: "Unique identifier of this assessment." })
	id: string;

	/** Identifier of the assignment that is being evaluated by this assessment. */
	@ApiProperty({
		description: "Identifier of the assignment that is being evaluated by this assessment."
	})
	assignmentId: string;

	assignment?: AssignmentDto;

	@ApiProperty({ description: "Marks the assessment as a draft." })
	isDraft: boolean;

	/** The amount of points that the student or group achieved with their submission. */
	@ApiProperty({
		description:
			"The amount of points that the student or group achieved with their submission."
	})
	achievedPoints?: number;

	/** A comment providing additional feedback. */
	@ApiPropertyOptional({ description: "A comment providing additional feedback." })
	comment?: string;

	/** If a group submission is being evaluated, contains the identifier of the group. */
	@ApiPropertyOptional({
		description:
			"If a group submission is being evaluated, contains the identifier of the group."
	})
	groupId?: string;

	/** The group, whose submission was evaluated by this assessment. */
	group?: GroupDto;

	/** If a single user is being evaluated, contains the identifier of the user. */
	@ApiPropertyOptional({
		description: "If a single user is being evaluated, contains the identifier of the user."
	})
	userId?: string;

	/** If assessment targets a single user, contains the user. */
	participant?: ParticipantDto;

	/** Identifier of the creator of this assessment. */
	@ApiPropertyOptional({ description: "Identifier of the creator of this assessment." })
	creatorId: string;

	/** The creator of this assessment. */
	creator?: UserDto;

	@ApiPropertyOptional({
		description: "Identifier of the last user that updated this assessment."
	})
	lastUpdatedById?: UserId;
	lastUpdatedBy?: UserDto;

	creationDate?: Date;
	updateDate?: Date;

	partialAssessments?: PartialAssessmentDto[];
}

/** Version of AssessmentDto containing only properties that can used for creation. */
export class AssessmentCreateDto extends OmitType(AssessmentDto, [
	"group",
	"participant",
	"assignment",
	"creator",
	"creatorId",
	"id",
	"lastUpdatedBy",
	"lastUpdatedById",
	"creationDate",
	"updateDate"
]) {}
/** Version of AssessmentDto containing only properties that can be updated. */
export class AssessmentUpdateDto {
	achievedPoints?: number;
	isDraft?: boolean;
	comment?: string;
	partialAssessments?: PartialAssessmentDto[];
}
