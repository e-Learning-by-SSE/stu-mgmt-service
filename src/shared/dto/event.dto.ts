export class EventDto {
	event: string;
	timestamp: Date;
	payload?: Record<string, unknown>;
}
