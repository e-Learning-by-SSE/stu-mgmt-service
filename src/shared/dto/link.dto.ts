export class LinkDto {
	url: string;
	name: string;
}
