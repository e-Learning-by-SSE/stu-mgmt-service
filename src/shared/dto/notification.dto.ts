import { AssignmentId } from "../../course/entities/assignment.entity";
import { CourseId } from "../../course/entities/course.entity";
import { GroupId } from "../../course/entities/group.entity";
import { Event } from "../../course/events/events";
import { UserId } from "../entities/user.entity";

export class NotificationDto {
	event: Event;
	courseId: CourseId;
	assignmentId?: AssignmentId;
	groupId?: GroupId;
	userId?: UserId;
	payload?: Record<string, unknown>;
}
