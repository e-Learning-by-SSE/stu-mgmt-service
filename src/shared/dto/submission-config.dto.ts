import { IsNotEmpty } from "class-validator";

export class SubmissionConfigDto {
	@IsNotEmpty()
    tool: string;
    @IsNotEmpty()
	config: string;
}
