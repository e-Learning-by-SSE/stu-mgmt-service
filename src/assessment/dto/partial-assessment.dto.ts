import { IsNotEmpty } from "class-validator";
import { MarkerDto } from "./marker.dto";

export class PartialAssessmentDto {
	/**
	 * Unique identifier for this partial assessment.
	 * Must be unique within a single assessment.
	 * If not specified, it will be automatically generated.
	 * @example "task_1"
	 */
	key?: string;
	@IsNotEmpty()
	title: string;
	/** Determines whether the partial assessment should be included, when the assessment is not a draft anymore. */
	draftOnly: boolean;
	comment?: string;
	points?: number;
	markers?: MarkerDto[];
}
