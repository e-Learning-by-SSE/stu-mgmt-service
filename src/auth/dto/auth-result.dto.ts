import { UserDto } from "../../shared/dto/user.dto";

export class AuthResultDto {
	user: UserDto;
	accessToken: string;
	expiration?: Date;
}
