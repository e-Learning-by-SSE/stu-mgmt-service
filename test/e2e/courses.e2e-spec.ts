import { getConnection } from "typeorm";
import { CourseCreateDto } from "../../src/course/dto/course/course-create.dto";
import { CourseFilter } from "../../src/course/dto/course/course-filter.dto";
import {
	COURSE_CONFIG_COURSE_INFO_2_2020,
	COURSE_CONFIG_JAVA_1920
} from "../mocks/course-config/course-config.mock";
import { CoursesMock, COURSE_INFO_2_2020, COURSE_JAVA_1920 } from "../mocks/courses.mock";
import { GROUP_1_JAVA } from "../mocks/groups/groups.mock";
import { USER_STUDENT_JAVA } from "../mocks/users.mock";
import { TestSetup } from "../utils/e2e";
import { copy } from "../utils/object-helper";

const course = COURSE_JAVA_1920; // the course that will be used for testing

describe("Courses E2E", () => {
	let setup: TestSetup;

	beforeAll(async () => {
		setup = await TestSetup.create();
	});

	afterAll(async () => {
		await setup.teardown();
	});

	describe("GET-REQUESTS of CourseController (e2e)", () => {
		beforeAll(async () => {
			await setup.clearDb();
			await setup.dbMockService.createAll();
		});

		it("(GET) /courses Retrieves all courses", () => {
			return setup
				.request()
				.get("/courses")
				.expect(({ body }) => {
					expect(body.length).toEqual(CoursesMock.length);
				});
		});

		it("(GET) /courses With filter -> Retrieves the filtered courses", () => {
			const filter: CourseFilter = {
				shortname: "java",
				title: "Programmier"
			};

			return setup
				.request()
				.get(`/courses?shortname=${filter.shortname}&title=${filter.title}`)
				.expect(({ body }) => {
					expect(body.length).toEqual(3);
				});
		});

		it("(GET) /courses With filter -> Retrieves all iterations of the course", () => {
			const filter: CourseFilter = {
				shortname: "java"
			};

			const expectedLength = CoursesMock.filter(c => c.shortname === filter.shortname).length;
			console.assert(
				expectedLength > 1,
				"The should be multiple course with the same shortname."
			);

			return setup
				.request()
				.get(`/courses?shortname=${filter.shortname}`)
				.expect(({ body }) => {
					expect(body.length).toEqual(expectedLength);
				});
		});

		it("(GET) /courses/{courseId} Retrieves the course", () => {
			return setup
				.request()
				.get(`/courses/${course.id}`)
				.expect(({ body }) => {
					expect(body.id).toEqual(course.id);
					expect(body).toMatchSnapshot();
				});
		});

		it("(GET) /courses/{name}/semester/{semester} Retrieves the course", () => {
			return setup
				.request()
				.get(`/courses/${course.shortname}/semester/${course.semester}`)
				.expect(({ body }) => {
					expect(body.id).toEqual(course.id);
				});
		});
	});

	describe("POST-REQUESTS of CourseController (empty db) (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
		});

		it("(POST) /courses Creates the given course and returns it", () => {
			const courseToCreate: CourseCreateDto = {
				...copy(course),
				config: COURSE_CONFIG_JAVA_1920
			};

			return setup
				.request()
				.post("/courses")
				.send(courseToCreate)
				.expect(201)
				.expect(({ body }) => {
					expect(body.shortname).toEqual(course.shortname);
				});
		});
	});

	describe("POST-REQUESTS for relations (db contains data) of CourseController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createCourses();
			await setup.dbMockService.createCourseConfig();
			await setup.dbMockService.createAdmissionCriteria();
			await setup.dbMockService.createGroupSettings();
			await setup.dbMockService.createUsers();
		});

		it("(POST) /courses/{courseId}/users/{userId} No password required -> Adds user to course", () => {
			const courseNoPassword = COURSE_INFO_2_2020;
			console.assert(
				COURSE_CONFIG_COURSE_INFO_2_2020.password == undefined,
				"Course password should be undefined"
			);
			const user = USER_STUDENT_JAVA;

			return setup
				.request()
				.post(`/courses/${courseNoPassword.id}/users/${user.id}`)
				.expect(201);
		});

		it("(POST) /courses/{courseId}/users/{userId} Correct password -> Adds user to course", () => {
			const user = USER_STUDENT_JAVA;

			return setup
				.request()
				.post(`/courses/${course.id}/users/${user.id}`)
				.send({ password: COURSE_CONFIG_JAVA_1920.password }) // the correct password
				.expect(201);
		});

		it("(POST) /courses/{courseId}/users/{userId} Incorrect password -> Throws BadRequestException", () => {
			const user = USER_STUDENT_JAVA;

			return setup
				.request()
				.post(`/courses/${course.id}/users/${user.id}`)
				.send({ password: "incorrect" }) // an incorrect password
				.expect(400);
		});

		it("(POST) /courses/{courseId}/groups Creates the given group and returns it (Part 1/2)", () => {
			const group = GROUP_1_JAVA;

			return setup
				.request()
				.post(`/courses/${COURSE_JAVA_1920.id}/groups`)
				.send(group)
				.expect(201)
				.expect(({ body }) => {
					expect(body.name).toEqual(group.name);
				});
		});
	});

	describe("DELETE-REQUESTS (Db contains data) of CourseController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createAll();
		});

		it("(DELETE) /courses/{courseId} Deletes the course", () => {
			return setup.request().delete(`/courses/${course.id}`).expect(200);
		});
	});
});
