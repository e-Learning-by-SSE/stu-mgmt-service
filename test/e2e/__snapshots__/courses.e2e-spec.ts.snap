// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Courses E2E GET-REQUESTS of CourseController (e2e) (GET) /courses/{courseId} Retrieves the course 1`] = `
Object {
  "admissionCriteria": Object {
    "rules": Array [
      Object {
        "achievedPercentRounding": Object {
          "type": "UP_NEAREST_INTEGER",
        },
        "assignmentType": "HOMEWORK",
        "requiredPercent": 50,
        "type": "REQUIRED_PERCENT_OVERALL",
      },
      Object {
        "achievedPercentRounding": Object {
          "decimals": 0,
          "type": "DECIMALS",
        },
        "assignmentType": "TESTAT",
        "requiredPercent": 50,
        "type": "REQUIRED_PERCENT_OVERALL",
      },
      Object {
        "achievedPercentRounding": Object {
          "type": "NONE",
        },
        "allowedFailures": 2,
        "assignmentType": "HOMEWORK",
        "requiredPercent": 50,
        "type": "INDIVIDUAL_PERCENT_WITH_ALLOWED_FAILURES",
      },
    ],
  },
  "groupSettings": Object {
    "allowGroups": true,
    "autoJoinGroupOnCourseJoined": true,
    "mergeGroupsOnAssignmentStarted": true,
    "selfmanaged": true,
    "sizeMax": 3,
    "sizeMin": 2,
  },
  "id": "java-wise1920",
  "isClosed": false,
  "links": Array [
    Object {
      "name": "Example URL",
      "url": "http://example-url.com",
    },
  ],
  "semester": "wise1920",
  "shortname": "java",
  "title": "Programmierpraktikum I: Java",
}
`;
